src/plugins/discover/public/application/main/components/top_nav/discover_topnav.tsx

```diff
@@ -193,11 +160,65 @@ export const DiscoverTopNav = ({
   );
 
   const { topNavBadges, topNavMenu } = useDiscoverTopNav({ stateContainer });
-  const topNavProps = !services.serverless && {
-    badges: topNavBadges,
-    config: topNavMenu,
-    setMenuMountPoint,
-  };
+  const setMenuMountPoint = getHeaderActionMenuMounter();
+  const topNavProps = useMemo(() => {
+    if (services.serverless) {
+      return undefined;
+    }
+
+    return {
+      badges: topNavBadges,
+      config: topNavMenu,
+      setMenuMountPoint,
+    };
+  }, [services.serverless, setMenuMountPoint, topNavBadges, topNavMenu]);
+
+  const dataViewPickerProps: DataViewPickerProps = useMemo(() => {
+    const isESQLModeEnabled = uiSettings.get(ENABLE_ESQL);
+    const supportedTextBasedLanguages: DataViewPickerProps['textBasedLanguages'] = isESQLModeEnabled
+      ? [TextBasedLanguages.ESQL]
+      : [];
+
+    return {
+      trigger: {
+        label: dataView?.getName() || '',
+        'data-test-subj': 'discover-dataView-switch-link',
+        title: dataView?.getIndexPattern() || '',
+      },
+      currentDataViewId: dataView?.id,
+      onAddField: addField,
+      onDataViewCreated: createNewDataView,
+      onCreateDefaultAdHocDataView: stateContainer.actions.createAndAppendAdHocDataView,
+      onChangeDataView: stateContainer.actions.onChangeDataView,
+      textBasedLanguages: supportedTextBasedLanguages,
+      shouldShowTextBasedLanguageTransitionModal: Boolean(
```

Comment:
Current implementation (with a boolean value) does not work correctly as it is not listening to changes in `savedSearchState`. We would need to move it outside of `dataViewPickerProps` useMemo.

![Feb-21-2024 11-40-58](https://github.com/elastic/kibana/assets/1415710/84e77f01-0115-4f81-8d0c-527fcc9c0b91)

Reply:
Yeah unfortunately I didn't realize this until after I pushed the changes... In any case, fixed it here: https://github.com/elastic/kibana/pull/177107/commits/856a0eb718b598bceeee0bdaf008d3f23081b2a6. For some reason my functional tests also weren't catching it due to the query I was using, so I updated them to use the same query you used which reproduced the issue and should catch any future regressions here.